man/texmath.1.md
---
title: texmath
section: 1
footer: texmath manual
date: July 22, 2014
...

# NAME

texmath - convert between math formats.

# SYNOPSIS

texmath [*options*] [file\*]

# DESCRIPTION

`texmath` acts as a pipe, reading from stdin or from the specified
files and writing to stdout.  It converts math in the specified input
format (see `-f` below) to math in the specified output format (see `-t`
below).

If `texmath` is invoked under the name `texmath-cgi` (via a symbolic
link, or through simple renaming), it will act as a CGI script.  It will
expect query parameters for `from`, `to`, `input`, and optionally
`inline`, and return a JSON object with either `error` and a message or
`success` and the converted result.

# OPTIONS

`-f` *FORMAT*
:   Specify input ("from") format:  `tex`, `mathml`, `omml`, `native`.
    Defaults to `tex`.

`-t` *FORMAT*
:   Specify output ("to") format:  `tex`, `omml`, `xhtml`, `mathml`,
    `typst`, `eqn`, `starmath`, `pandoc`, `native`.  Defaults to `mathml`.

`--inline`
:   Use the inline display style.

`-h, --help`
:   Print usage information.

`-V, --version`
:   Print version.

# AUTHORS

John MacFarlane, Matthew Pickering, Jesse Rosenthal.

# SEE ALSO

The `texmath` source code and all documentation may be downloaded
from <http://github.com/jgm/texmath/>.


changelog
texmath (0.12.8.13)

  * Remove special override for `\perp` in Text.TeXMath.Readers.TeX.Commands
    (#247). This caused `\perp` to be read as U+22A5 instead of U+27C2.  This
    addresses the mismatch with the TeX writer (which associates `\bot` with
    U+22A5 and `\perp` with U+27C2).

  * Typst writer:

    + Fix several issues with accents and attachments (#245).
    + Fix handling of some EOver with combining accents (#245).
    + Escape backslash in text context (#245).

texmath (0.12.8.12)

  * TeX writer: render prime and superscripted prime as `'` (#246).

  * TeX reader:

    + Don't crash on array with `\hline` before blank cell (#244).
    + Skip whitespace in array column specifier (#244).

  * OMML writer:

    + Fix order of dPr attributes (#243).

  * Typst writer:

    + Escape commas (#242). Otherwise we can get bad results e.g. in fractions,
      when the commas separate arguments.

  * Require typst-symbols 0.1.7, update tests.

texmath (0.12.8.11)

  * TeX reader: Ignore `@{..}` and `!{..}` in array alignment specifiers (#241).

  * TeX reader: ignore `\color` instead of crashing (#225).

texmath (0.12.8.10)

  * TeX reader: allow `\lVert .. \vVert` to create an EDelimited (#238).

  * Typst writer: improved handling of primes (#239).
    Use `'` instead of e.g. `prime`. Don't put a space before primes.

  * Typst writer: improve rendering of EDelimited (#238).

  * Typst writer: use `mid()` for middle delimiters (#238).

texmath (0.12.8.9)

  * Parse TeX `\mathbf` as both bold and upright (#236).

texmath (0.12.8.8)

  * TeX reader: support unicode-math Greek symbols, e.g. `\Alpha` (#235).
    This includes symbols like `\Alpha` and `\omicron` that weren't
    defined in original TeX.

  * Use typst-symbols 0.1.6

texmath (0.12.8.7)

  * TeX reader: convert Bin symbols to Ord when appropriate (#234).
    E.g. in '-3', we should have an Ord rather than a Bin, so
    the spacing will be appropriate.

  * Pandoc writer: fix spacing inside EDelimited (#234).
    Previously spaces around binary operators were omitted when
    they occurred inside parens or brackets.

  * test-texmath: allow pandoc output.

texmath (0.12.8.6)

  * Typst writer: avoid redundant `lr`s (#233).

texmath (0.12.8.5)

  * Typst writer: use ASCII symbols when possible instead of symbols (#232).
    E.g., `+` instead of `plus`. Add `\` to characters needing escape.
    Enhance list of characters that need escaping.

  * Typst writer: fixed EBoxed output so it includes a border.

  * Handle `\ddot` better in conversion to typst (#231).

  * Use typst-symbols 0.1.5

texmath (0.12.8.4)

  * TeX reader: ignore `\allowbreak` (#230).

  * TeX reader: handle `*{5}{lr}` in array column specifier (#229).

  * OMML reader: allow `m:e` to be missing in `m:nary` (#228).
    Technically this is not allowed, according to the spec, but
    Word and LibreOffice seem to tolerate it.

texmath (0.12.8.3)

  * OMML writer: use "on" and "off" instead of "1" and "0" for
    m:CT_OnOff type.  It is said that "1" and "0" work in Word
    but not Powerpoint.

texmath (0.12.8.2)

  * Typst writer: use binom instead of a fraction (jgm/pandoc#9063).

texmath (0.12.8.1)

  * Typst writer: several fixes (#223, Lleu Yang).

    + Escape quotes (") in inQuotes
    + Accent `\8407` corresponds to `arrow()`
    + Write `#none`'s for matrices with blanks at the beginning of a row

texmath (0.12.8)

  * Expose Text.TeXMath.Shared [API change]

  * Typst writer: Fix bug where 's' turned into 'space' (#219).

  * Typst writer: Fix handling of overline (#214).

  * Typst writer: Fix underbrace (#217).

  * Typst writer: Improve some accents (#216).

  * TeX writer: don't include \\ on last line of matrix.

  * TeX writer: Remove escaping of spaces inside \text{}.
    It isn't needed, and it causes problems in MathJax rendering.

  * TeX reader: allow empty matrices.

  * MathML writer: Fix rendering of vectors (#218).

  * Depend on external typst-symbols package.

texmath (0.12.7.1)

 * Typst writer:

   + Improve under/overbrace/bracket/line.
   + Fix bugs with super/subscript grouping (#212).
   + Fix case where super/subscript is on an empty element,
     by inserting a zws.

texmath (0.12.7)

  * Add typst writer. New module: Text.TeXMath.Writers.Typst.

  * TeX reader: Support multilined environment. Closes #210.

texmath (0.12.6)

  * MathML writer:

    + Use style with CSS as well as columnalign (#205).
      This seems to be needed by browser implementations of MathML.
    + Remove reliance on mstyle (#205). mstyle doesn't seem to be
      supported any more, at least in browser implementations of MathML,
      and the documentation indicates that it is treated like mrow now
      that styles can go directly on child elements. This commit removes
      our use of mstyle. Instead of using mstyle, we change mathvariant
      attributes on descendent elements (and displaystyle attributes on
      direct children, in the case of fractions).
    + Extend our existing use of unicode replacements, since many
      implementations don't properly handle mathvariant. We now get
      variant characters for mo, mn, and all elements that can sensibly
      take them, not just mi and mtext.
    + Omit mathvariant attribute unless we can't find appropriate Unicode.
      When MathML is displayed by ODT, having BOTH a bold math Unicode
      character and a mathvariant="bold" attribute seems to confuse it.
      (Browsers don't care either way.) This gives us more compact and
      readable output, as well.

texmath (0.12.5.5)

  * Allow pandoc-types 1.23.

  * TeX reader: remove false positives for isConvertible (#204).
    "Convertible" symbols are those in which subscripts render
    under the symbol in display environments, and as subscripts
    in inline environments. Previously the TeX parser recognized
    all relation and binary symbols as convertible, which does not
    match TeX's behavior.


extra/texmath.hs
{-# LANGUAGE CPP #-}
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Text.TeXMath
import Data.Char (isSpace)
import Text.XML.Light
import System.IO
import System.Environment
import System.Console.GetOpt
import System.Exit
import Data.Maybe
import Text.Pandoc.Definition
import Network.URI (unEscapeString)
import Data.Aeson (encode, (.=), object)
import qualified Data.ByteString.Lazy as B
import qualified Data.Text as T
import qualified Data.Text.IO as T
import Data.Version ( showVersion )
import Text.Show.Pretty (ppShow)
import Paths_texmath (version)

tshow :: Show a => a -> T.Text
tshow = T.pack . ppShow

inHtml :: Element -> Element
inHtml e =
  add_attr (Attr (unqual "xmlns") "http://www.w3.org/1999/xhtml") $
  unode "html"
    [ unode "head" $
        add_attr (Attr (unqual "content") "application/xhtml+xml; charset=UTF-8") $
        add_attr (Attr (unqual "http-equiv") "Content-Type") $
        unode "meta" ()
    , unode "body" e ]

type Reader = T.Text -> Either T.Text [Exp]
data Writer = XMLWriter (DisplayType -> [Exp] -> Element)
            | StringWriter (DisplayType -> [Exp] -> T.Text)
            | PandocWriter (DisplayType -> [Exp] -> Maybe [Inline])

readers :: [(T.Text, Reader)]
readers = [
    ("tex", readTeX)
  , ("typst", readTypst)
  , ("mathml", readMathML)
  , ("omml", readOMML)
  , ("native", readNative)]

readNative :: T.Text -> Either T.Text [Exp]
readNative s =
  case reads (T.unpack s) of
       ((exps, ws):_) | all isSpace ws -> Right exps
       ((_, (_:_)):_) -> Left "Could not read whole input as native Exp"
       _ -> Left "Could not read input as native Exp"

writers :: [(T.Text, Writer)]
writers = [
    ("native", StringWriter (\_ es -> tshow es) )
  , ("tex", StringWriter (\_ -> writeTeX))
  , ("eqn", StringWriter writeEqn)
  , ("typst", StringWriter writeTypst)
  , ("omml",  XMLWriter writeOMML)
  , ("xhtml",   XMLWriter (\dt e -> inHtml (writeMathML dt e)))
  , ("mathml",   XMLWriter writeMathML)
  , ("pandoc", PandocWriter writePandoc)]

data Options = Options {
    optDisplay :: DisplayType
  , optIn :: T.Text
  , optOut :: T.Text }

def :: Options
def = Options DisplayBlock "tex" "mathml"

options :: [OptDescr (Options -> IO Options)]
options =
  [ Option [] ["inline"]
      (NoArg (\opts -> return opts {optDisplay = DisplayInline}))
      "Use the inline display style"
  , Option "f" ["from"]
      (ReqArg (\s opts -> return opts {optIn = T.pack s}) "FORMAT")
      ("Input format: " <> T.unpack (T.intercalate ", " (map fst readers)))
  , Option "t" ["to"]
      (ReqArg (\s opts -> return opts {optOut = T.pack s}) "FORMAT")
      ("Output format: " <> T.unpack (T.intercalate ", " (map fst writers)))
   , Option "v" ["version"]
      (NoArg (\_ -> do
                      hPutStrLn stderr $ "Version " <> showVersion version
                      exitWith ExitSuccess))
      "Print version"

    , Option "h" ["help"]
        (NoArg (\_ -> do
                        prg <- getProgName
                        hPutStrLn stderr (usageInfo (prg <> " [OPTIONS] [FILE*]") options)
                        exitWith ExitSuccess))
      "Show help"
  ]

output :: DisplayType -> Writer -> [Exp] -> T.Text
output dt (XMLWriter w) es    = output dt (StringWriter (\dt' -> T.pack . ppElement . w dt' )) es
output dt (StringWriter w) es = w dt es
output dt (PandocWriter w) es = tshow (fromMaybe fallback (w dt es))
  where fallback = [Math mt $ writeTeX es]
        mt = case dt of
                  DisplayBlock  -> DisplayMath
                  DisplayInline -> InlineMath

err :: Bool -> Int -> T.Text -> IO a
err cgi code msg = do
  if cgi
     then B.putStr $ encode $ object ["error" .= msg]
     else T.hPutStrLn stderr msg
  exitWith $ ExitFailure code

ensureFinalNewline :: T.Text -> T.Text
ensureFinalNewline xs = case T.unsnoc xs of
  Nothing        -> xs
  Just (_, '\n') -> xs
  _              -> xs <> "\n"

urlUnencode :: T.Text -> T.Text
urlUnencode = T.pack . unEscapeString . plusToSpace . T.unpack
  where plusToSpace ('+':xs) = "%20" <> plusToSpace xs
        plusToSpace (x:xs)   = x : plusToSpace xs
        plusToSpace []       = []

main :: IO ()
main = do
  progname <- getProgName
  let cgi = progname == "texmath-cgi"
  if cgi
     then runCGI
     else runCommandLine

runCommandLine :: IO ()
runCommandLine = do
  args <- getArgs
  let (actions, files, _) = getOpt RequireOrder options args
  opts <- foldl (>>=) (return def) actions
  inp <- case files of
              [] -> T.getContents
              _  -> T.concat <$> mapM T.readFile files
  reader <- case lookup (optIn opts) readers of
                  Just r -> return r
                  Nothing -> err False 3 "Unrecognised reader"
  writer <- case lookup (optOut opts) writers of
                  Just w -> return w
                  Nothing -> err False 5 "Unrecognised writer"
  case reader inp of
        Left msg -> err False 1 msg
        Right v  -> T.putStr $ ensureFinalNewline
                             $ output (optDisplay opts) writer v
  exitWith ExitSuccess

runCGI :: IO ()
runCGI = do
  query <- T.getContents
  let topairs xs = case T.break (=='=') xs of
                     (ys, zs) -> case T.uncons zs of
                       Just ('=', zs') -> (urlUnencode ys, urlUnencode zs')
                       _               -> (urlUnencode ys, "")
  let pairs = map topairs $ T.split (== '&') query
  inp <- case lookup "input" pairs of
                 Just x  -> return x
                 Nothing -> err True 3 "Query missing 'input'"
  reader <- case lookup "from" pairs of
                 Just x  -> case lookup x readers of
                                 Just y  -> return y
                                 Nothing -> err True 5 "Unsupported value of 'from'"
                 Nothing -> err True 3 "Query missing 'from'"
  writer <- case lookup "to" pairs of
                 Just x  -> case lookup x writers of
                                 Just y  -> return y
                                 Nothing -> err True 7 "Unsupported value of 'to'"
                 Nothing -> err True 3 "Query missing 'to'"
  let inline = isJust $ lookup "inline" pairs
  putStr "Content-type: text/json; charset=UTF-8\n\n"
  case reader inp of
        Left msg -> err True 1 msg
        Right v  -> B.putStr $ encode $ object
                       [ "success" .=
                       ensureFinalNewline (output
                        (if inline
                            then DisplayInline
                            else DisplayBlock) writer v) ]
  exitWith ExitSuccess
